### ISSUE TITLE
To be done

### ISSUE THREAD

**Issue by aboeglin** [open-sorcerers/madlib#8]
## Features:
- [x] Make inference work for recursive functions
- [x] Pattern match **List**
- [ ] Pattern match **String**
- [ ] Error reporting and formatting : **WIP**
- [x] Support Arrow args for ADTs ( `data A = A (Num -> Num)` )
- [x] Record deconstruction ( Patterns )
- [x] Record spread operator
- [x] String literals character, replace double quotes with single quotes or back ticks ( preference seems to be for back ticks )
- [x] Rename `switch` / `case` ( `where` / `is` ? )
- [ ] Wrap up compilation of patterns and check ADT types correctly in generated JS code ( using `__buildCtorParam` )
- [x] Define and implement a first CLI : **WIP**
- [x] Add operators `&&` and `||`, `>`, `>=`, `<`, `=<`, `!`
- [x] Add missing operator: `-` ( unary ), ...
- [ ] Verify and complete places where returns are allowed within expressions
- [x] Tuples solving/compiling
- [x] Tuple pattern matching
- [x] Tuple typing
- [x] Update comments and use `//` instead of `—`
- [x] Make curlies optional ( where, if, more ? )
- [x] Add ternary expression grammar
- [x] type aliases
- [x] bundle compiler option that generates a single js output for browsers : **WIP** with rollup for now
- [ ] Prelude basics
- [x] ADTs should also be exportable and importable ( Also that should go hand in hand with the point below on checking ADT usages )

## Code:
- [ ] `__buildCtorParam` should be put in a separate `.mjs` file that is imported where needed and not directly embedded in the generated js.
- [ ] Rework InferError and either split errors ( Infer, Import, xyz ), or keep them unified but then rename them from InferError to simple Error ?
- [x] We need to add some checks on ADT definitions and usage ( Mostly that names being used actually exist ( usage ), or that they don't ( definition ) )
- [ ] Improve tests and test a lot more scenarios
- [x] Consider removing Records for **Exp** as we never have more than 2 or 3 args anyways
- [x] Move pos and type information outside of Exp, so that Exp is cleaner and only concerned about the grammar
- [x] Create a Typed and/or Located type that contains type or location information
- [x] Update grammar so that a Typing, can be defined in two ways:
  - ```javascript
      // Must come right before the function definition, and we just make it a TypedExp Assignment (=) Abstraction x (x+1)
      inc :: Num -> Num
      inc = (x) => x + 1
      ```
  - ```javascript
      // Directly attached to an expression and we make it a TypedExp App (App (+) (1)) 4
      4 + 1 :: Num
      ```
  That way we can more easily connect the type annotation to the wrong function in case of mismatches and report more precise errors, and it would also make the AST clearer as for now in the case of functions we just assign the TypedExp to a var to be looked up in the env. We could then potentially also remove that entry in the Env ( `envtypings :: Typings` ).
- [x] Production for TypedExp for abstractions is too confusing and won't work well the way it's done. So we need to split the type and the abstraction to make the grammar simpler, and make it a `NamedTypedExp` ( that assigns to a name instead of an Exp ), and make the match during desugaring step ( AST.Source -> AST.Canonical )

## CI:
- [ ] Run tests before publishing a release !

## Legal:
- [ ] Pick and add licence info

**Comment by brekk** [comment]
I think we're not quite done with everything here, but soon we should cut it off and make new issues, as this issue is rather large.

**Comment by aboeglin** [comment]
Yeah, also some will probably never be done like returns in expressions. I think that's more an organic thing that will evolve. So we should select the bullets that have a known scope, and put them in their own issue and close this one then ?

**Comment by aboeglin** [comment]
@brekk 
I would just close that for now and eventually take some of it like error reporting into specific tickets / issues ?

**Comment by brekk** [comment]
@aboeglin Sounds good to me.
